File: pedsim_gazebo_plugin/scripts/spawn_pedsim_agents.py
#!/usr/bin/env python3
"""
@mantainer: stephenadhi

"""
import functools
import rclpy
from ament_index_python.packages import get_package_share_directory
from gazebo_msgs.srv import SpawnEntity, SetEntityState
from geometry_msgs.msg import Pose
from pedsim_msgs.msg import AgentStates
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data


class AgentSpawner(Node):

    def __init__(self):
        super().__init__('agent_spawner')
        qos_profile = qos_profile_sensor_data
        self.sub = self.create_subscription(AgentStates, 
                                            'pedsim_simulator/simulated_agents',
                                            self.actor_poses_callback, qos_profile)
        self.sub  # prevent unused variable warning
        pedsim_dir = get_package_share_directory('pedsim_gazebo_plugin')
        file_xml = open(pedsim_dir + "/models/person_standing/model.sdf")
        self.xml_string = file_xml.read()

        self.spawn_client = self.create_client(SpawnEntity, 'spawn_entity')
        self.set_state_client = self.create_client(SetEntityState, '/gazebo_spawner/set_entity_state')
        
        while not self.spawn_client.wait_for_service(timeout_sec=1.0):
            self.get_logger().warn('service not available, waiting again...')
        # self.get_logger().info("service: spawn_sdf_model is available ....")

        max_ped = 200
        self.is_spawned = [False] * max_ped

    def actor_poses_callback(self, actors):
        for idx, actor in enumerate(actors.agent_states):
            actor_id = str(actor.id)

            if not self.is_spawned[idx]:
                spawn_future = self.spawn_entity(actor_id=actor_id, model_pose=actor.pose)
                if spawn_future:
                    callback = functools.partial(self.callback, actor_id, actor.pose, idx)
                    spawn_future.add_done_callback(callback)
            else:
                self.set_entity_state(actor_id=actor_id, model_pose=actor.pose)

    def callback(self, actor_id, model_pose, idx, future):
        if future.exception() is None:
            self.is_spawned[idx] = True
            self.set_entity_state(actor_id=actor_id, model_pose=model_pose)

    def spawn_entity(self, actor_id, model_pose):
        try:
            req = SpawnEntity.Request()
            req.name = actor_id
            req.xml = self.xml_string
            req.robot_namespace = ""
            req.initial_pose = model_pose
            req.reference_frame = "world"
            self.spawn_client.call_async(req)
            return self.spawn_client.call_async(req)  # return the Future object
        except Exception as e:
            self.get_logger().warn(f"Failed to spawn entity: {e}")
            return None

    def set_entity_state(self, actor_id, model_pose):
        try:
            req = SetEntityState.Request()
            req.state.name = actor_id
            req.state.pose = model_pose
            self.set_state_client.call_async(req)
        except Exception as e:
            self.get_logger().warn(f"Failed to update agent state: {e}")

def main(args=None):
    rclpy.init(args=args)
    agent_spawner = AgentSpawner()

    try:
        rclpy.spin(agent_spawner)
    except KeyboardInterrupt:
        pass

    agent_spawner.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: pedsim_gazebo_plugin/src/spawn_pedsim_agents.cpp
#include "rclcpp/rclcpp.hpp"
#include "gazebo_msgs/srv/spawn_entity.hpp"
#include "gazebo_msgs/srv/set_entity_state.hpp"
#include "pedsim_msgs/msg/agent_states.hpp"
#include "ament_index_cpp/get_package_share_directory.hpp"

#include <fstream>
#include <vector>

class AgentSpawner : public rclcpp::Node
{
public:
    AgentSpawner()
        : Node("agent_spawner")
    {
        auto qos = rclcpp::SensorDataQoS();
        subscription_ = this->create_subscription<pedsim_msgs::msg::AgentStates>(
            "pedsim_simulator/simulated_agents", qos,
            std::bind(&AgentSpawner::actor_poses_callback, this, std::placeholders::_1));

        std::string pedsim_dir = ament_index_cpp::get_package_share_directory("pedsim_gazebo_plugin");
        std::ifstream file_xml(pedsim_dir + "/models/person_standing/model.sdf");
        if (file_xml.is_open()) {
        xml_string_ = std::string((std::istreambuf_iterator<char>(file_xml)),
                                std::istreambuf_iterator<char>());
        } else {
            RCLCPP_WARN(this->get_logger(), "Failed to open file at path: %s", pedsim_dir.c_str());
        }


        spawn_client_ = this->create_client<gazebo_msgs::srv::SpawnEntity>("spawn_entity");
        set_state_client_ = this->create_client<gazebo_msgs::srv::SetEntityState>("/gazebo_spawner/set_entity_state");

        while (!spawn_client_->wait_for_service(std::chrono::seconds(1))) {
            RCLCPP_WARN(this->get_logger(), "service not available, waiting again...");
        }

        is_spawned_.resize(max_ped_, false);
    }

private:
    void actor_poses_callback(const pedsim_msgs::msg::AgentStates::SharedPtr actors)
    {
        for (size_t idx = 0; idx < actors->agent_states.size(); ++idx)
        {
            auto& actor = actors->agent_states[idx];
            std::string actor_id = std::to_string(actor.id);

            if (!is_spawned_[idx])
            {
                auto spawn_future = spawn_entity(actor_id, actor.pose);
                futures_.push_back(spawn_future); // Save the future
                size_t future_idx = futures_.size() - 1; // Get the index of the saved future
    
                // Start a new thread to wait for the future
                std::thread([this, actor_id, model_pose=actor.pose, idx, future_idx]() {
                    auto& future = futures_[future_idx]; // Access the future by index
                    future.wait(); // Wait for the future to be ready
                    callback(actor_id, model_pose, idx, future_idx); // Call the callback
                }).detach();
            } else {
                set_entity_state(actor_id, actor.pose);
            }
        }
    }

    void callback(std::string actor_id, geometry_msgs::msg::Pose model_pose, int idx, size_t future_idx)
    {
        auto& future = futures_[future_idx]; // Access the future by index
        auto response = future.get();
        if (response && response->success) {
            is_spawned_[idx] = true;
            set_entity_state(actor_id, model_pose);
        } else {
            RCLCPP_WARN(this->get_logger(), "Failed to spawn entity: %s", actor_id.c_str());
        }
    }

    rclcpp::Client<gazebo_msgs::srv::SpawnEntity>::SharedFuture spawn_entity(
        std::string actor_id, geometry_msgs::msg::Pose model_pose)
    {
        auto request = std::make_shared<gazebo_msgs::srv::SpawnEntity::Request>();
        request->name = actor_id;
        request->xml = xml_string_;
        request->robot_namespace = "";
        request->initial_pose = model_pose;
        request->reference_frame = "world";
        auto future = spawn_client_->async_send_request(request);
        return future;
    }

    void set_entity_state(std::string actor_id, geometry_msgs::msg::Pose model_pose)
    {
        auto request = std::make_shared<gazebo_msgs::srv::SetEntityState::Request>();
        request->state.name = actor_id;
        request->state.pose = model_pose;
        set_state_client_->async_send_request(request);
    }

    std::vector<rclcpp::Client<gazebo_msgs::srv::SpawnEntity>::SharedFuture> futures_;
    rclcpp::Subscription<pedsim_msgs::msg::AgentStates>::SharedPtr subscription_;
    rclcpp::Client<gazebo_msgs::srv::SpawnEntity>::SharedPtr spawn_client_;
    rclcpp::Client<gazebo_msgs::srv::SetEntityState>::SharedPtr set_state_client_;
    std::string xml_string_;
    std::vector<bool> is_spawned_;
    const int max_ped_ = 200;
};

int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<AgentSpawner>());
    rclcpp::shutdown();
    return 0;
}